dex-it/src/test/scala/com/wavesplatform/it/sync/MarketOrderSuite.scala

```diff
@@ -0,0 +1,380 @@
+        accountBalanceWBefore + marketOrderAmount - anotherOrderAmount - fee * (marketOrderAmount - anotherOrderAmount) / marketOrderAmount)
+    }
+  }
+
+  "Validation market orders" - {
+    val price = 0.4.usd
+
+    "should be accepted if there is no way to fill it completely (sum of all orders in order book < amount of market order" in {
+      placeOrders(bob, wavesUsdPair, OrderType.SELL)(100.waves -> price)
+
+      var total = 0L
+      node
+        .orderBook(wavesUsdPair)
+        .bids
+        .foreach(order => {
+          total += order.amount
+        })
+
+      node.waitOrderProcessed(wavesUsdPair,
+                              node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, OrderType.BUY, total + 1000.waves, price, fee)).message.id)
+    }
+
+    "should be accepted if user doesn't have enough Waves to pay fee, but he take waves from order result " in {
+      val amount   = 100.waves
+      val price    = 0.1.usd
+      val transfer = 100.usd
+
+      val account = createAccountWithBalance() {
+        transfer -> Some(UsdId.toString)
+      }
+
+      placeOrders(alice, wavesUsdPair, OrderType.SELL)(
+        amount -> price
+      )
+
+      node.waitOrderProcessed(wavesUsdPair,
+                              node.placeMarketOrder(node.prepareOrder(account, wavesUsdPair, OrderType.BUY, amount, price, fee)).message.id)
+
+      node.accountBalances(account.toAddress.toString)._1 should be(amount - fee)
+    }
+
+    "should be rejected if the price is too low for completely filling by current opened orders (BUY)" in {
+      val amount      = 100.waves
+      val marketPrice = price - 0.1.usd
+
+      placeOrders(bob, wavesUsdPair, OrderType.SELL)(amount -> price)
+
+      assertBadRequestAndMessage(node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, OrderType.BUY, amount, marketPrice, fee)),
+                                 tooLowPrice("buy", "0.3"))
+    }
+
+    "should be rejected if the price is too high for completely filling by current opened orders (SELL)" in {
+      val amount      = 100.waves
+      val marketPrice = price + 0.1.usd
+
+      placeOrders(bob, wavesUsdPair, OrderType.BUY)(amount -> price)
+
+      assertBadRequestAndMessage(node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, OrderType.SELL, amount, marketPrice, fee)),
+                                 tooHighPrice("sell", "0.5"))
+    }
+
+    "should be rejected if amount of the buy market order more then user could buy" ignore /* because of TODO: DEX-457 */ {
+      val amount   = 101.waves
+      val price    = 1.1.usd
+      val transfer = 100.usd
+
+      val account = createAccountWithBalance() {
+        transfer -> Some(UsdId.toString)
+      }
+
+      placeOrders(alice, wavesUsdPair, OrderType.SELL)(
+        amount -> price
+      )
+
+      val orderBook = node.orderBook(wavesUsdPair)
+      orderBook.bids shouldNot be(empty)
+      orderBook.asks should be(empty)
```

Comment:
Asks should be nonempty whereas bids should be empty, solve this after DEX-457

Reply:
ok, after 457 I will refactor this test